lib/espree.js

```diff
@@ -0,0 +1,314 @@
+ * @param {int} start The index at which the comment starts.
+ * @param {int} end The index at which the comment ends.
+ * @param {Location} startLoc The location at which the comment starts.
+ * @param {Location} endLoc The location at which the comment ends.
+ * @returns {Object} The comment object.
+ * @private
+ */
+function convertAcornCommentToEsprimaComment(block, text, start, end, startLoc, endLoc) {
+    const comment = {
+        type: block ? "Block" : "Line",
+        value: text
+    };
+
+    if (typeof start === "number") {
+        comment.start = start;
+        comment.end = end;
+        comment.range = [start, end];
+    }
+
+    if (typeof startLoc === "object") {
+        comment.loc = {
+            start: startLoc,
+            end: endLoc
+        };
+    }
+
+    return comment;
+}
+
+module.exports = () => Parser => class Espree extends Parser {
+    constructor(options, code) {
+        if (typeof options !== "object" || options === null) {
+            options = {};
+        }
+        if (typeof code !== "string" && !(code instanceof String)) {
+            code = String(code);
+        }
+
+        const ecmaFeatures = options.ecmaFeatures || {};
+        const ecmaVersion = normalizeEcmaVersion(options.ecmaVersion);
+        const isModule = options.sourceType === "module";
+        const tokenTranslator =
+            options.tokens === true ?
+                new TokenTranslator(tokTypes, code) :
+                null;
+
+        // Initialize acorn parser.
+        super({
+            ecmaVersion: isModule ? Math.max(6, ecmaVersion) : ecmaVersion,
+            sourceType: isModule ? "module" : "script",
+            ranges: options.range === true || options.attachComment === true,
+            locations: options.loc === true,
+
+            // Truthy value is true for backward compatibility.
+            allowReturnOutsideFunction: Boolean(ecmaFeatures.globalReturn),
+
+            // Collect tokens
+            onToken: (token) => {
+                if (tokenTranslator) {
+                    // Use `tokens`, `ecmaVersion`, and `jsxAttrValueToken` in the state.
+                    tokenTranslator.onToken(token, this[STATE]);
+                }
+                if (token.type !== tokTypes.eof) {
+                    this[STATE].lastToken = token;
+                }
+            },
+
+            // Collect comments
+            onComment: (block, text, start, end, startLoc, endLoc) => {
+                if (this[STATE].comments) {
+                    const comment = convertAcornCommentToEsprimaComment(block, text, start, end, startLoc, endLoc);
+                    this[STATE].comments.push(comment);
+
+                    if (options.attachComment === true) {
+                        commentAttachment.addComment(comment);
+                    }
+                }
+            }
+        }, code);
+
+        // TODO: remove global state.
+        commentAttachment.reset();
```

Comment:
At this point, since we don't use comment attachment in ESLint, I wonder if we should actually remove the comment attachment option in the future? I don't think it adds much value as of right now.

Reply:
I agree. We can remove the comment attachment logic in the future.
But this PR focuses on upgrading acorn since it's a breaking change that removes `attarchComment` option.

Reply:
Yes, sorry I wasn't clear. Wasn't trying to suggest we change that in this PR!